parsons/freshdesk/freshdesk.py

```diff
@@ -0,0 +1,178 @@
+from parsons.utilities import check_env
+from parsons.utilities.api_connector import APIConnector
+import re
+from parsons import Table
+import logging
+
+logger = logging.getLogger(__name__)
+
+PAGE_SIZE = 100
+
+
+class Freshdesk():
+    """
+    Instantiate FreshDesk Class
+
+    `Args:`
+        domain:
+            The subdomain of the FreshDesk account. Not required if ``FRESHDESK_DOMAIN``
+            env variable set.
+        api_key:
+            The FreshDesk provided application key. Not required if ``FRESHDESK_API_KEY``
+            env variable set.
+    `Returns:`
+        FreshDesk Class
+    """
+
+    def __init__(self, domain, api_key):
+
+        self.api_key = check_env.check('FRESHDESK_API_KEY', api_key)
+        self.domain = check_env.check('FRESHDESK_DOMAIN', domain)
+        self.uri = f'https://{self.domain}.freshdesk.com/api/v2/'
+        self.client = APIConnector(self.uri, auth=(self.api_key, 'x'))
+
+    def get_request(self, endpoint, params=None, **kwargs):
+        # Internal method to make a get request.
+
+        base_params = {'per_page': PAGE_SIZE}
+
+        if params:
+            base_params.update(params)
+
+        r = self.client.request(self.uri + endpoint, 'GET', params=base_params)
+        self.client.validate_response(r)
+        data = r.json()
+
+        # Paginate
+        while 'link' in r.headers.keys():
+            logger.info(f'Retrieving another page of {PAGE_SIZE} records.')
+            url = re.search('<(.*)>', r.headers['link']).group(1)
+            r = self.client.request(url, 'GET', params=params)
+            self.client.validate_response(r)
+            data.extend(r.json())
+
+        return data
+
+    def transform_table(self, tbl, expand_custom_fields=None):
+        # Internal method to transform a table prior to returning
+
+        tbl.move_column('id', 0)
+        tbl.sort()
+        if expand_custom_fields:
+            tbl.unpack_dict('custom_fields', prepend=False)
```

Comment:
Is there already an intended use case for unpacking in this flat manner? I don't think there's any reason to hold up the PR over design considerations here, but Eliot had mentioned in a call that perhaps there might be unintended consequences to breaking each dictionary field into a new column when custom fields might change over time. Should we consider doing some checking here or somewhere else to make sure schemas are aligned when the custom columns change?

Reply:
I think that is should be checked, but I would suggest that it be checked as part of a synchronization script that leverages these methods. The use case is for users that don't anticipate their custom fields changing over time and want to 1) avoid additional side tables and joins in a database 2) parsing of nested jsons in a DB.

I set the default to not unpack, so that the data model remains consistent.